### PR TITLE
Social panel twitter api

### PR DESCRIPTION
**Pull request by ikantoch**
In the last commit, I moved the access token from request to the variable. First, as we spoke last time I have created the initialize method. However, at work, I often use constants and I find this very useful. It is because in case I need to see the token I can simply call the class name and constant name like `TwitterClient::ACCESS_TOKEN`. I'm not sure which take is better. Tests are passing when using `@access_token` and const `ACCESS_TOKEN`.

### CODE REVIEW

**Comment by marekciupak** [comment]
> In the last commit, I moved the access token from request to the variable. First, as we spoke last time I have created the initialize method. However, at work, I often use constants and I find this very useful. It is because in case I need to see the token I can simply call the class name and constant name like TwitterClient::ACCESS_TOKEN. I'm not sure which take is better. Tests are passing when using @access_token and const ACCESS_TOKEN.

Dobra rozkmina 👍 

> I'm not sure which take is better.

Let's talk about it tomorrow.

File: social_panel/Gemfile.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+source 'https://rubygems.org'
+
+gem "faraday", "~> 1.0"
+
+gem "rspec", "~> 3.9"
+
+gem "json", "~> 2.3"
+
+gem "webmock", "~>3.8"
```
Comment (marekciupak) [inline social_panel/Gemfile.rb]: I'd suggest changing the config of your editor ;) It should be adding missing lines at the end of the files for you ;)

File: social_panel/lib/twitter_client.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+require 'faraday'
+require 'JSON'
+require 'pry'
+
+class TwitterClient
+  URL = 'https://api.twitter.com/1.1'
+  ACCESS_TOKEN = "Bearer #{ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN']}"
```
Comment (marekciupak) [inline social_panel/lib/twitter_client.rb]: Naming:

I'd say `ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN']` is the access token and `"Bearer #{ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN']}"` is the `AUTHORIZATION_HEADER`.

File: social_panel/spec/integration_tests/search_by_phrase_spec.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+require './spec/spec_helper.rb'
+require './lib/twitter_client.rb'
+WebMock.allow_net_connect!
+
+RSpec.describe 'finding tweet phrase' do
```
Comment (marekciupak) [inline social_panel/spec/integration_tests/search_by_phrase_spec.rb]:
```suggestion
RSpec.describe 'finding tweets by phrase' do
```

File: social_panel/spec/unit_tests/twitter_client_spec.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    stubbed_body =  "{\"created_at\":\"Mon Apr 06 10:11:36 +0000 2020\",\"id\":1247104664144936960,\"id_str\":\"1247104664144936960\",\"text\":\"My first twitter*\"}"
+
+    stub_request(:get, "https://api.twitter.com/1.1/statuses/show.json?id=1247104664144936960").
+      with(
+        headers: {
+          'Accept'=>'*/*',
+          'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
+          'Authorization'=>"Bearer #{ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN']}"
+        }).
```
Comment (marekciupak) [inline social_panel/spec/unit_tests/twitter_client_spec.rb]: I'd make sure that `Authorization` has a correct value and I wouldn't test the other 2 headers.

```suggestion
        headers: {'Authorization'=>"Bearer #{ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN']}"}).
```

File: social_panel/spec/unit_tests/twitter_client_spec.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      to_return(status: 200, body: stubbed_body, headers: {})
+
+    result = TwitterClient.new.get_tweet_by_id('1247104664144936960')
+    expect(result).to include(
+                        "created_at" => "Mon Apr 06 10:11:36 +0000 2020",
+                        "text" => "My first twitter*"
+                      )
+  end
```
Comment (marekciupak) [inline social_panel/spec/unit_tests/twitter_client_spec.rb:26]: The first test looks good. Where is the test to `search_tweets_by_phrase` method? ;)))

File: social_panel/lib/twitter_client.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+  # def initialize(access_token: ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN'])
+  #   @access_token = "Bearer #{access_token}"
+  # end
+
+  def get_tweet_by_id(tweet_id)
+    response = Faraday.get("#{URL}/statuses/show.json?id=#{tweet_id}") do |request|
+      request.headers['Authorization'] = ACCESS_TOKEN
+    end
```
Comment (marekciupak) [inline social_panel/lib/twitter_client.rb]: As a next step of your refactor: try to use this syntax:

```ruby
url = 'http://sushi.com/nigiri/sake.json'
resp = Faraday.get(url, {a: 1}, {'Accept' => 'application/json'})
# => GET http://sushi.com/nigiri/sake.json?a=1
```
I copied it from the doc: https://lostisland.github.io/faraday/usage/

Note that you will be refactoring here internal things in this class. You can rely on your tests. The tests should still pass.

File: social_panel/lib/twitter_client.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  # URL = 'https://api.twitter.com/1.1'
+  # TOKEN = "Bearer #{ENV.fetch('SOCIAL_PANEL_TWITTER_ACCESS_TOKEN')}"
+
+  def connection
+      Faraday.new(
+        url: 'https://api.twitter.com/1.1',
+        headers: { 'Authorization' => "Bearer #{ENV['SOCIAL_PANEL_TWITTER_ACCESS_TOKEN']}" },
+      )
+  end
```
Comment (marekciupak) [inline social_panel/lib/twitter_client.rb:17]: Really nice idea @ikantoch 🥇

Good job.

File: social_panel/lib/twitter_client.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+  def get_tweet_by_id(tweet_id)
+    response = connection.get("statuses/show.json?id=#{tweet_id}")
```
Comment (marekciupak) [inline social_panel/lib/twitter_client.rb:20]: Looks much better!

You can still do one improvement. The documentations shows this example:

```ruby
url = 'http://sushi.com/nigiri/sake.json'
resp = Faraday.get(url, {a: 1}, {'Accept' => 'application/json'})
# => GET http://sushi.com/nigiri/sake.json?a=1
```

https://lostisland.github.io/faraday/usage/

You can get rid of this `?id=#{tweet_id}` part and use the syntax from that example. Try it ;)

File: social_panel/lib/twitter_client.rb (hunk 7 above)
Comment (marekciupak) [reply social_panel/lib/twitter_client.rb:17]: There is one interesting implication of this. For example when I run this code:

```ruby
get_tweet_by_id(1)
get_tweet_by_id(2)
get_tweet_by_id(3)
```

the `Faraday.new` will be executed 3 times and we can think about it - is it good or not, etc.

We will talk about it tomorrow :)

File: social_panel/spec/unit_tests/twitter_client_spec.rb (hunk 5 above)
Comment (ikantoch) [reply social_panel/spec/unit_tests/twitter_client_spec.rb:26]: I've never created one, only integration test. Will do!

**Pull request merged**